In [2]:
import os 
import numpy as np
import pandas as pd

from sklearn import tree

In [3]:
!pip install sklearn

  Running setup.py bdist_wheel for sklearn: started
  Running setup.py bdist_wheel for sklearn: finished with status 'done'
  Stored in directory: C:\Users\DELL\AppData\Local\pip\Cache\wheels\76\03\bb\589d421d27431bcd2c6da284d5f2286c8e3b2ea3cf1594c074
Successfully built sklearn


In [4]:
# Using Decision tree as a non-paramteric method to create heuristic rules for classification and regression

In [5]:
#benefits of using a decision tree
#1. simple to understand and interpret the results
#2. Tree structure, which represents the rules can be visualized
#3. less data manipulation, transformation or normalization
#4. less costly algorithm
#5. handle both numerical and categorical data
#6. multinomial problem: multiclass can be handles easily
#7. validate the model easily

In [6]:
-(.38)*np.log2(0.38)+(-(0.62)*np.log2(0.62)) #base entropy 0.95

0.9580420222262995

In [7]:
np.log2(1.0)

0.0

In [8]:
#for branch where emabrek = C
-(.55)*np.log2(0.55)+(-(0.45)*np.log2(0.45))
#for branch where emabrek = S
-(.43)*np.log2(0.43)+(-(0.57)*np.log2(0.57))
#for branch where emabrek = Q
-(.43)*np.log2(0.43)+(-(0.57)*np.log2(0.57))
#for branch where emabrek = B
0
weithed_entropy = ((168/884)*(-(.55)*np.log2(0.55)+(-(0.45)*np.log2(0.45)))) + ((70/884) * (-(.43)*np.log2(0.43)+(-(0.57)*np.log2(0.57)))) + ((644/884)*(-(.43)*np.log2(0.43)+(-(0.57)*np.log2(0.57))))
weithed_entropy

0.9849072905154983

In [9]:
F1= 0.2
F2 = -0.2
F3 = 0.4

In [10]:
# Maximum variance
# Gini Coefficient
# Entropy


In [11]:
Informaion_gain = (-(.38)*np.log2(0.38)+(-(0.62)*np.log2(0.62))) - weithed_entropy
Informaion_gain

-0.026865268289198774

In [12]:
-(.333)*np.log2(0.333)+(-(0.333)*np.log2(0.333))+(-(0.333)*np.log2(0.333)) #base entropy

1.584819511673234

In [13]:
# How to decide, which variable to be placed at the root node
# Ans: compare the information gain by putting each variable at the root node
# the variable that gives maximum information gain need to be selected

# Information Gain = Entropy at base - Weighted Entropy at the node after the split

# How to select the brances

# How long you can expand the decision tree (max_depth)
# Ans:
# conditions to stop the decision tree
# if all the samples in a node belong to same class
# if all the features are already part of the tree
# if all the samples entered the tree model


# what is stopping criteria to terminate a tree
# if all the samples in a node belong to same class
# if all the features are already part of the tree
# if all the samples entered the tree model

# if i expand the tree, what are the limitations
# model ovefit, train accuracy will be different from test accuracy

# how to identify right fit for the tree
# a balance between train and test

# how to prune the tree
# put conditions, min sample, max_depth, min leaf, delta chnage in IG

# validation of a tree
# k-fold cross validation can be used to prune the decision tree

In [14]:
from sklearn.datasets import load_iris
iris = load_iris()

In [15]:
iris.target

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

In [16]:
print (iris.DESCR)

.. _iris_dataset:

Iris plants dataset
--------------------

**Data Set Characteristics:**

    :Number of Instances: 150 (50 in each of three classes)
    :Number of Attributes: 4 numeric, predictive attributes and the class
    :Attribute Information:
        - sepal length in cm
        - sepal width in cm
        - petal length in cm
        - petal width in cm
        - class:
                - Iris-Setosa
                - Iris-Versicolour
                - Iris-Virginica
                
    :Summary Statistics:

    ============== ==== ==== ======= ===== ====================
                    Min  Max   Mean    SD   Class Correlation
    ============== ==== ==== ======= ===== ====================
    sepal length:   4.3  7.9   5.84   0.83    0.7826
    sepal width:    2.0  4.4   3.05   0.43   -0.4194
    petal length:   1.0  6.9   3.76   1.76    0.9490  (high!)
    petal width:    0.1  2.5   1.20   0.76    0.9565  (high!)
    ============== ==== ==== ======= ===== ===========

In [89]:
tree.DecisionTreeClassifier() # tree initializations

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
                       max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort=False,
                       random_state=None, splitter='best')

In [90]:
tree1 = tree.DecisionTreeClassifier()
tree1.fit(iris.data,iris.target)

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
                       max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort=False,
                       random_state=None, splitter='best')

In [91]:
tree1.score(iris.data,iris.target)

1.0

In [92]:
iris.feature_names

['sepal length (cm)',
 'sepal width (cm)',
 'petal length (cm)',
 'petal width (cm)']

In [93]:
tree1.feature_importances_

array([0.        , 0.01333333, 0.06405596, 0.92261071])

In [94]:
13/40

0.325

In [95]:
#tree pruning conditions

#1. if all the samples in a leaf node belong to one specific class
#2. if all the samples absorbed in the model
#3. if all the variables are entered into the model

In [96]:
# classification criteria

#1. Gini Impurity Index : Sum(Pmk*(1-Pmk))
# m = 3 (Iris Dataset)
# 1,2,3.......K-1 number of levels in a target variable
# k= number of samples for each class


In [97]:
#2. Cross Entropy: Sum(-Pmk*log(Pmk))
# m = 3 (Iris Dataset)
# 1,2,3.......K-1 number of levels in a target variable
# k= number of samples for each class


In [98]:
#3. Misclassification Approach: 1-max(Pmk)
# m = 3 (Iris Dataset)
# 1,2,3.......K-1 number of levels in a target variable
# k= number of samples for each class


In [99]:
# n = 150
# Setosa = 50
# P(setosa,samples) = 50/150 

In [100]:
df = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/wine/wine.data',header=None,sep=',')

In [101]:
df.head()

0      1     2     3     4    5     6     7     8     9     10    11    12  \
0   1  14.23  1.71  2.43  15.6  127  2.80  3.06  0.28  2.29  5.64  1.04  3.92   
1   1  13.20  1.78  2.14  11.2  100  2.65  2.76  0.26  1.28  4.38  1.05  3.40   
2   1  13.16  2.36  2.67  18.6  101  2.80  3.24  0.30  2.81  5.68  1.03  3.17   
3   1  14.37  1.95  2.50  16.8  113  3.85  3.49  0.24  2.18  7.80  0.86  3.45   
4   1  13.24  2.59  2.87  21.0  118  2.80  2.69  0.39  1.82  4.32  1.04  2.93   

     13  
0  1065  
1  1050  
2  1185  
3  1480  
4   735

In [102]:
df.columns = ['Target','Alcohol','Malic acid','Ash','Alcalinity of ash ','Magnesium',' Total phenols',
             'Flavanoids','Nonflavanoid phenols','Proanthocyanins','Color intensity','Hue','diluted wines','Proline']

In [103]:
df.head()

Target  Alcohol  Malic acid   Ash  Alcalinity of ash   Magnesium  \
0       1    14.23        1.71  2.43                15.6        127   
1       1    13.20        1.78  2.14                11.2        100   
2       1    13.16        2.36  2.67                18.6        101   
3       1    14.37        1.95  2.50                16.8        113   
4       1    13.24        2.59  2.87                21.0        118   

    Total phenols  Flavanoids  Nonflavanoid phenols  Proanthocyanins  \
0            2.80        3.06                  0.28             2.29   
1            2.65        2.76                  0.26             1.28   
2            2.80        3.24                  0.30             2.81   
3            3.85        3.49                  0.24             2.18   
4            2.80        2.69                  0.39             1.82   

   Color intensity   Hue  diluted wines  Proline  
0             5.64  1.04           3.92     1065  
1             4.38  1.05           3.40     1050  
2             5.68  1.03           3.17     1185  
3             7.80  0.86           3.45     1480  
4             4.32  1.04           2.93      735

1) Alcohol
 	2) Malic acid
 	3) Ash
	4) Alcalinity of ash  
 	5) Magnesium
	6) Total phenols
 	7) Flavanoids
 	8) Nonflavanoid phenols
 	9) Proanthocyanins
	10)Color intensity
 	11)Hue
 	12)OD280/OD315 of diluted wines
 	13)Proline     

In [104]:
from sklearn.datasets import load_iris
from sklearn import tree

In [105]:
iris = load_iris()

In [106]:
iris.data.shape # input data

(150, 4)

In [107]:
#print iris.DESCR

In [108]:
iris.feature_names # getting the feature names

['sepal length (cm)',
 'sepal width (cm)',
 'petal length (cm)',
 'petal width (cm)']

In [109]:
iris.target.shape # target variable or dependent variable

(150,)

In [110]:
mytree = tree.DecisionTreeClassifier()
mytree

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
                       max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort=False,
                       random_state=None, splitter='best')

In [111]:
mytree = tree.DecisionTreeClassifier(criterion='entropy')
mytree

DecisionTreeClassifier(class_weight=None, criterion='entropy', max_depth=None,
                       max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort=False,
                       random_state=None, splitter='best')

In [112]:
dt1 = mytree.fit(iris.data, iris.target)
dt1

DecisionTreeClassifier(class_weight=None, criterion='entropy', max_depth=None,
                       max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort=False,
                       random_state=None, splitter='best')

In [113]:
dt1.feature_importances_

array([0.0115876 , 0.0115876 , 0.06606582, 0.91075898])

In [114]:
dt1.fit(iris.data, iris.target)
tree.export_graphviz(dt1,out_file='None')  

In [115]:
tree.export_graphviz(dt1,out_file='tree.dot')
dotfile = open("tree.dot", 'w')
tree.export_graphviz(dt1, out_file = dotfile, feature_names = iris.feature_names)
dotfile.close()

In [116]:
dot_d = tree.export_graphviz(dt1,out_file='Decision Tree')  

In [117]:
from sklearn.datasets import load_iris
from sklearn import tree
iris = load_iris()
clf = tree.DecisionTreeClassifier()
clf = clf.fit(iris.data, iris.target)

In [118]:
with open("iris.dot", 'w') as f:
    f = tree.export_graphviz(clf, out_file=f)

In [119]:
import os

In [120]:
os.unlink('iris.dot')

In [121]:
from IPython.display import Image

In [122]:
print(dot_d)

None


In [123]:
dot_data = tree.export_graphviz(clf, out_file='Decision Tree', 
                         feature_names=iris.feature_names,  
                         class_names=iris.target_names,  
                         filled=True, rounded=True,  
                         special_characters=True)  


In [124]:
!pip install pydotplus

You are using pip version 9.0.1, however version 19.1.1 is available.
You should consider upgrading via the 'python -m pip install --upgrade pip' command.


In [125]:
#graph = pydotplus.graph_from_dot_data(dot_data)  
#graph

In [126]:
import pydotplus

In [127]:
!pip install pydotplus

You are using pip version 9.0.1, however version 19.1.1 is available.
You should consider upgrading via the 'python -m pip install --upgrade pip' command.


In [128]:
!pip install msgpack

You are using pip version 9.0.1, however version 19.1.1 is available.
You should consider upgrading via the 'python -m pip install --upgrade pip' command.


In [129]:
# three methods determine the importance of features
# 1. mean decrease in gini coefficient
# 2. informtion gain -cross entropy
# 3. variance based approach

In [130]:
dt1.score(iris.data,iris.target)

1.0

In [134]:
# decision tree for regression based problems
from sklearn.datasets import load_boston
from sklearn.model_selection import cross_val_score
from sklearn.tree import DecisionTreeRegressor


In [135]:
boston = load_boston()
regressor = DecisionTreeRegressor(random_state=0)

In [138]:
from sklearn.model_selection import train_test_split

In [139]:
X_train,X_test,Y_train,Y_test = train_test_split(boston.data, boston.target,test_size=0.20,random_state=1234)

In [140]:
X_train.shape,Y_train.shape,X_test.shape,Y_test.shape

((404, 13), (404,), (102, 13), (102,))

In [141]:
regressor.fit(X_train,Y_train)

DecisionTreeRegressor(criterion='mse', max_depth=None, max_features=None,
                      max_leaf_nodes=None, min_impurity_decrease=0.0,
                      min_impurity_split=None, min_samples_leaf=1,
                      min_samples_split=2, min_weight_fraction_leaf=0.0,
                      presort=False, random_state=0, splitter='best')

In [142]:
help(DecisionTreeRegressor())

Help on DecisionTreeRegressor in module sklearn.tree.tree object:

class DecisionTreeRegressor(BaseDecisionTree, sklearn.base.RegressorMixin)
 |  A decision tree regressor.
 |  
 |  Read more in the :ref:`User Guide <tree>`.
 |  
 |  Parameters
 |  ----------
 |  criterion : string, optional (default="mse")
 |      The function to measure the quality of a split. Supported criteria
 |      are "mse" for the mean squared error, which is equal to variance
 |      reduction as feature selection criterion and minimizes the L2 loss
 |      using the mean of each terminal node, "friedman_mse", which uses mean
 |      squared error with Friedman's improvement score for potential splits,
 |      and "mae" for the mean absolute error, which minimizes the L1 loss
 |      using the median of each terminal node.
 |  
 |      .. versionadded:: 0.18
 |         Mean Absolute Error (MAE) criterion.
 |  
 |  splitter : string, optional (default="best")
 |      The strategy used to choose the split at ea

In [143]:
regressor.score(X_train,Y_train) #training accuracy

1.0

In [144]:
regressor.score(X_test,Y_test) #test accuracy

0.8878863675631422

In [145]:
from sklearn.model_selection import GridSearchCV
from time import time
from operator import itemgetter

In [151]:
# create a function to perform grid search with cross validation and returns the accuracy values
def GridSearch_BestParam(X, y, clf, param_grid,cv=10):
    grid_search = GridSearchCV(clf,
                              param_grid=param_grid,
                              cv=cv)
    start= time()
    grid_search.fit(X,y)
    top_params=grid_search.best_estimator_
    return top_params

In [152]:
param_grid = {"criterion": ["mae","mse","friedman_mse"],
             "min_samples_split": [2,10,20],
             "max_depth": [2,5,10],
             "min_samples_leaf":[1,2,10]}

In [153]:
cvDT = tree.DecisionTreeRegressor()

In [154]:
top_para = GridSearch_BestParam(X_train,Y_train, cvDT, param_grid, cv=10)
print (top_para)

DecisionTreeRegressor(criterion='mae', max_depth=10, max_features=None,
                      max_leaf_nodes=None, min_impurity_decrease=0.0,
                      min_impurity_split=None, min_samples_leaf=1,
                      min_samples_split=20, min_weight_fraction_leaf=0.0,
                      presort=False, random_state=None, splitter='best')


E:\New folder\lib\site-packages\sklearn\model_selection\_search.py:813: DeprecationWarning: The default of the `iid` parameter will change from True to False in version 0.22 and will be removed in 0.24. This will change numeric results when test-set sizes are unequal.
  DeprecationWarning)


In [157]:
top_para

DecisionTreeRegressor(criterion='mae', max_depth=10, max_features=None,
                      max_leaf_nodes=None, min_impurity_decrease=0.0,
                      min_impurity_split=None, min_samples_leaf=1,
                      min_samples_split=20, min_weight_fraction_leaf=0.0,
                      presort=False, random_state=None, splitter='best')

In [ ]:
paramCV = top_score[0].parameters
paramCV

In [ ]:
best_model = tree.DecisionTreeRegressor(max_depth=paramCV['max_depth'], min_samples_split=paramCV['min_samples_split'],
                                    min_samples_leaf=paramCV['min_samples_leaf'],
                                   criterion=paramCV['criterion'])
best_model

In [160]:
best_model = top_para.fit(X_train,Y_train)

In [161]:
YValid_Pred = best_model.predict(X_test) # use XTest, XValid to generate prediction
YValid_Pred

array([33.2 , 23.8 ,  8.8 , 21.7 , 15.15, 23.8 , 19.95, 17.2 , 17.55,
       27.25, 17.2 , 21.7 , 21.7 , 18.2 , 17.55, 24.6 ,  8.8 , 19.  ,
       19.95,  8.8 , 32.55, 20.6 , 19.95, 14.4 , 23.8 , 20.6 , 19.95,
       24.6 , 27.1 , 17.2 , 19.95, 19.  , 18.2 , 11.95, 24.6 , 14.1 ,
       20.  , 21.7 ,  8.8 ,  8.8 ,  8.8 , 18.2 , 42.8 , 27.1 , 19.95,
       27.25, 27.25, 50.  , 19.  , 24.6 , 19.95, 33.2 , 28.7 , 24.6 ,
       50.  , 32.55,  8.8 , 33.2 , 28.7 , 21.7 , 32.55, 15.15, 19.95,
       32.55,  8.8 , 20.6 , 33.2 , 20.  , 50.  , 42.8 , 15.  , 23.8 ,
       17.2 , 17.2 , 24.6 , 15.15, 42.8 , 22.45, 17.2 , 17.2 , 42.8 ,
       50.  , 24.6 , 17.2 , 20.6 , 17.2 , 23.8 , 32.55, 20.6 , 20.6 ,
       19.  ,  7.55, 23.8 , 19.  , 13.8 , 42.8 , 19.  , 16.5 , 27.25,
       27.25, 19.95, 19.95])

In [162]:
from sklearn import metrics
metrics.r2_score(Y_test,YValid_Pred)

0.8556796648378939

In [163]:
best_model.score(X_train,Y_train)

0.9300850149395175

In [164]:
best_model.score(X_test,Y_test)

0.8556796648378939

In [165]:
from sklearn.metrics import confusion_matrix, classification_report


In [166]:
best_model.feature_importances_

array([0.05899659, 0.        , 0.02813048, 0.        , 0.0295241 ,
       0.43300299, 0.02678848, 0.06844224, 0.        , 0.01001342,
       0.02482709, 0.02162692, 0.29864767])

In [167]:
# overfiting the model
# drawback of using train/test model evaluation is incorrect
# to control over fitting of the model, we are going to use CV
# cross validation tunes the parameters, selects best model, does feature selection

In [168]:
data = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/wine-quality/winequality-red.csv',
                  sep=";")

In [169]:
data.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  
0      9.4        5  
1      9.8        5  
2      9.8        5  
3      9.8        6  
4      9.4        5

In [170]:
print(data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1599 entries, 0 to 1598
Data columns (total 12 columns):
fixed acidity           1599 non-null float64
volatile acidity        1599 non-null float64
citric acid             1599 non-null float64
residual sugar          1599 non-null float64
chlorides               1599 non-null float64
free sulfur dioxide     1599 non-null float64
total sulfur dioxide    1599 non-null float64
density                 1599 non-null float64
pH                      1599 non-null float64
sulphates               1599 non-null float64
alcohol                 1599 non-null float64
quality                 1599 non-null int64
dtypes: float64(11), int64(1)
memory usage: 150.0 KB
None


In [171]:
set(data.quality)

{3, 4, 5, 6, 7, 8}

In [172]:
data['quality'] = pd.Categorical(data.quality)

In [173]:
print(data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1599 entries, 0 to 1598
Data columns (total 12 columns):
fixed acidity           1599 non-null float64
volatile acidity        1599 non-null float64
citric acid             1599 non-null float64
residual sugar          1599 non-null float64
chlorides               1599 non-null float64
free sulfur dioxide     1599 non-null float64
total sulfur dioxide    1599 non-null float64
density                 1599 non-null float64
pH                      1599 non-null float64
sulphates               1599 non-null float64
alcohol                 1599 non-null float64
quality                 1599 non-null category
dtypes: category(1), float64(11)
memory usage: 139.3 KB
None


In [174]:
data.columns

Index(['fixed acidity', 'volatile acidity', 'citric acid', 'residual sugar',
       'chlorides', 'free sulfur dioxide', 'total sulfur dioxide', 'density',
       'pH', 'sulphates', 'alcohol', 'quality'],
      dtype='object')

In [175]:
df = data[['fixed acidity', 'volatile acidity', 'citric acid', 'residual sugar',
       'chlorides', 'free sulfur dioxide', 'total sulfur dioxide', 'density',
       'pH', 'sulphates', 'alcohol']]
target = data['quality']

In [176]:
X_train, X_test, y_train,y_test = train_test_split(data,target,test_size=0.10,random_state=1234)

In [177]:
X_train.shape, X_test.shape, y_train.shape,y_test.shape 

((1439, 12), (160, 12), (1439,), (160,))

In [178]:
from sklearn.tree import DecisionTreeClassifier
dt2 = DecisionTreeClassifier()

In [179]:
dt2

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
                       max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort=False,
                       random_state=None, splitter='best')

In [180]:
dt2.fit(X_train,y_train)

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
                       max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort=False,
                       random_state=None, splitter='best')

In [181]:
dt2.score(X_train,y_train)

1.0

In [182]:
dt2.score(X_test,y_test)

1.0

In [183]:
from sklearn.metrics import classification_report,confusion_matrix

In [184]:
pred_y = dt2.predict(X_test)

In [185]:
confusion_matrix(y_test,pred_y)

array([[ 3,  0,  0,  0,  0],
       [ 0, 71,  0,  0,  0],
       [ 0,  0, 61,  0,  0],
       [ 0,  0,  0, 23,  0],
       [ 0,  0,  0,  0,  2]], dtype=int64)

In [186]:
ad_data = pd.read_csv("https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data",sep=',',header=None)

| This data was extracted from the census bureau database found at
| http://www.census.gov/ftp/pub/DES/www/welcome.html
| Donor: Ronny Kohavi and Barry Becker,
|        Data Mining and Visualization
|        Silicon Graphics.
|        e-mail: ronnyk@sgi.com for questions.
| Split into train-test using MLC++ GenCVFiles (2/3, 1/3 random).
| 48842 instances, mix of continuous and discrete    (train=32561, test=16281)
| 45222 if instances with unknown values are removed (train=30162, test=15060)
| Duplicate or conflicting instances : 6
| Class probabilities for adult.all file
| Probability for the label '>50K'  : 23.93% / 24.78% (without unknowns)
| Probability for the label '<=50K' : 76.07% / 75.22% (without unknowns)
|
| Extraction was done by Barry Becker from the 1994 Census database.  A set of
|   reasonably clean records was extracted using the following conditions:
|   ((AAGE>16) && (AGI>100) && (AFNLWGT>1)&& (HRSWK>0))
|
| Prediction task is to determine whether a person makes over 50K
| a year.
|
| First cited in:
| @inproceedings{kohavi-nbtree,
|    author={Ron Kohavi},
|    title={Scaling Up the Accuracy of Naive-Bayes Classifiers: a
|           Decision-Tree Hybrid},
|    booktitle={Proceedings of the Second International Conference on
|               Knowledge Discovery and Data Mining},
|    year = 1996,
|    pages={to appear}}
|
| Error Accuracy reported as follows, after removal of unknowns from
|    train/test sets):
|    C4.5       : 84.46+-0.30
|    Naive-Bayes: 83.88+-0.30
|    NBTree     : 85.90+-0.28
|
|
| Following algorithms were later run with the following error rates,
|    all after removal of unknowns and using the original train/test split.
|    All these numbers are straight runs using MLC++ with default values.
|
|    Algorithm               Error
| -- ----------------        -----
| 1  C4.5                    15.54
| 2  C4.5-auto               14.46
| 3  C4.5 rules              14.94
| 4  Voted ID3 (0.6)         15.64
| 5  Voted ID3 (0.8)         16.47
| 6  T2                      16.84
| 7  1R                      19.54
| 8  NBTree                  14.10
| 9  CN2                     16.00
| 10 HOODG                   14.82
| 11 FSS Naive Bayes         14.05
| 12 IDTM (Decision table)   14.46
| 13 Naive-Bayes             16.12
| 14 Nearest-neighbor (1)    21.42
| 15 Nearest-neighbor (3)    20.35
| 16 OC1                     15.04
| 17 Pebls                   Crashed.  Unknown why (bounds WERE increased)
|
| Conversion of original data as follows:
| 1. Discretized agrossincome into two ranges with threshold 50,000.
| 2. Convert U.S. to US to avoid periods.
| 3. Convert Unknown to "?"
| 4. Run MLC++ GenCVFiles to generate data,test.
|
| Description of fnlwgt (final weight)
|
| The weights on the CPS files are controlled to independent estimates of the
| civilian noninstitutional population of the US.  These are prepared monthly
| for us by Population Division here at the Census Bureau.  We use 3 sets of
| controls.
|  These are:
|          1.  A single cell estimate of the population 16+ for each state.
|          2.  Controls for Hispanic Origin by age and sex.
|          3.  Controls by Race, age and sex.
|
| We use all three sets of controls in our weighting program and "rake" through
| them 6 times so that by the end we come back to all the controls we used.
|
| The term estimate refers to population totals derived from CPS by creating
| "weighted tallies" of any specified socio-economic characteristics of the
| population.
|
| People with similar demographic characteristics should have
| similar weights.  There is one important caveat to remember
| about this statement.  That is that since the CPS sample is
| actually a collection of 51 state samples, each with its own
| probability of selection, the statement only applies within
| state.


>50K, <=50K.

age: continuous.
workclass: Private, Self-emp-not-inc, Self-emp-inc, Federal-gov, Local-gov, State-gov, Without-pay, Never-worked.
fnlwgt: continuous.
education: Bachelors, Some-college, 11th, HS-grad, Prof-school, Assoc-acdm, Assoc-voc, 9th, 7th-8th, 12th, Masters, 1st-4th, 10th, Doctorate, 5th-6th, Preschool.
education-num: continuous.
marital-status: Married-civ-spouse, Divorced, Never-married, Separated, Widowed, Married-spouse-absent, Married-AF-spouse.
occupation: Tech-support, Craft-repair, Other-service, Sales, Exec-managerial, Prof-specialty, Handlers-cleaners, Machine-op-inspct, Adm-clerical, Farming-fishing, Transport-moving, Priv-house-serv, Protective-serv, Armed-Forces.
relationship: Wife, Own-child, Husband, Not-in-family, Other-relative, Unmarried.
race: White, Asian-Pac-Islander, Amer-Indian-Eskimo, Other, Black.
sex: Female, Male.
capital-gain: continuous.
capital-loss: continuous.
hours-per-week: continuous.
native-country: United-States, Cambodia, England, Puerto-Rico, Canada, Germany, Outlying-US(Guam-USVI-etc), India, Japan, Greece, South, China, Cuba, Iran, Honduras, Philippines, Italy, Poland, Jamaica, Vietnam, Mexico, Portugal, Ireland, France, Dominican-Republic, Laos, Ecuador, Taiwan, Haiti, Columbia, Hungary, Guatemala, Nicaragua, Scotland, Thailand, Yugoslavia, El-Salvador, Trinadad&Tobago, Peru, Hong, Holand-Netherlands.


In [187]:
ad_data.head()

0                  1       2           3   4                    5   \
0  39          State-gov   77516   Bachelors  13        Never-married   
1  50   Self-emp-not-inc   83311   Bachelors  13   Married-civ-spouse   
2  38            Private  215646     HS-grad   9             Divorced   
3  53            Private  234721        11th   7   Married-civ-spouse   
4  28            Private  338409   Bachelors  13   Married-civ-spouse   

                   6               7       8        9     10  11  12  \
0        Adm-clerical   Not-in-family   White     Male  2174   0  40   
1     Exec-managerial         Husband   White     Male     0   0  13   
2   Handlers-cleaners   Not-in-family   White     Male     0   0  40   
3   Handlers-cleaners         Husband   Black     Male     0   0  40   
4      Prof-specialty            Wife   Black   Female     0   0  40   

               13      14  
0   United-States   <=50K  
1   United-States   <=50K  
2   United-States   <=50K  
3   United-States   <=50K  
4            Cuba   <=50K

In [188]:
ad_data.columns = ['age','workclass','fnlwgt','education','education_num','marital_status','occupation',
                  'relationship','race','sex','capital_gain','capital_loss','hours_per_week','native_country','income']

In [189]:
ad_data.head()

age          workclass  fnlwgt   education  education_num  \
0   39          State-gov   77516   Bachelors             13   
1   50   Self-emp-not-inc   83311   Bachelors             13   
2   38            Private  215646     HS-grad              9   
3   53            Private  234721        11th              7   
4   28            Private  338409   Bachelors             13   

        marital_status          occupation    relationship    race      sex  \
0        Never-married        Adm-clerical   Not-in-family   White     Male   
1   Married-civ-spouse     Exec-managerial         Husband   White     Male   
2             Divorced   Handlers-cleaners   Not-in-family   White     Male   
3   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
4   Married-civ-spouse      Prof-specialty            Wife   Black   Female   

   capital_gain  capital_loss  hours_per_week  native_country  income  
0          2174             0              40   United-States   <=50K  
1             0             0              13   United-States   <=50K  
2             0             0              40   United-States   <=50K  
3             0             0              40   United-States   <=50K  
4             0             0              40            Cuba   <=50K

In [190]:
ad_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
age               32561 non-null int64
workclass         32561 non-null object
fnlwgt            32561 non-null int64
education         32561 non-null object
education_num     32561 non-null int64
marital_status    32561 non-null object
occupation        32561 non-null object
relationship      32561 non-null object
race              32561 non-null object
sex               32561 non-null object
capital_gain      32561 non-null int64
capital_loss      32561 non-null int64
hours_per_week    32561 non-null int64
native_country    32561 non-null object
income            32561 non-null object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


In [191]:
features = ad_data[['age','fnlwgt','education_num',
                  'capital_gain','capital_loss','hours_per_week']]
target = ad_data.income

In [192]:
male = pd.get_dummies(ad_data.sex,prefix='sex',drop_first=True)
workclass = pd.get_dummies(ad_data.workclass,prefix='workclass',drop_first=True)
education = pd.get_dummies(ad_data.education,prefix='education',drop_first=True)
marital_status = pd.get_dummies(ad_data.marital_status,prefix='marital_status',drop_first=True)
occupation = pd.get_dummies(ad_data.occupation,prefix='occupation',drop_first=True)
relationship = pd.get_dummies(ad_data.relationship,prefix='relationship',drop_first=True)
race = pd.get_dummies(ad_data.race,prefix='race',drop_first=True)

cdata = pd.concat([male,workclass,education,marital_status,occupation,relationship,race,features],axis=1)

In [193]:
cdata.head()

sex_ Male  workclass_ Federal-gov  workclass_ Local-gov  \
0          1                       0                     0   
1          1                       0                     0   
2          1                       0                     0   
3          1                       0                     0   
4          0                       0                     0   

   workclass_ Never-worked  workclass_ Private  workclass_ Self-emp-inc  \
0                        0                   0                        0   
1                        0                   0                        0   
2                        0                   1                        0   
3                        0                   1                        0   
4                        0                   1                        0   

   workclass_ Self-emp-not-inc  workclass_ State-gov  workclass_ Without-pay  \
0                            0                     1                       0   
1                            1                     0                       0   
2                            0                     0                       0   
3                            0                     0                       0   
4                            0                     0                       0   

   education_ 11th  ...  race_ Asian-Pac-Islander  race_ Black  race_ Other  \
0                0  ...                         0            0            0   
1                0  ...                         0            0            0   
2                0  ...                         0            0            0   
3                1  ...                         0            1            0   
4                0  ...                         0            1            0   

   race_ White  age  fnlwgt  education_num  capital_gain  capital_loss  \
0            1   39   77516             13          2174             0   
1            1   50   83311             13             0             0   
2            1   38  215646              9             0             0   
3            0   53  234721              7             0             0   
4            0   28  338409             13             0             0   

   hours_per_week  
0              40  
1              13  
2              40  
3              40  
4              40  

[5 rows x 59 columns]

In [194]:
cdata.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 59 columns):
sex_ Male                                32561 non-null uint8
workclass_ Federal-gov                   32561 non-null uint8
workclass_ Local-gov                     32561 non-null uint8
workclass_ Never-worked                  32561 non-null uint8
workclass_ Private                       32561 non-null uint8
workclass_ Self-emp-inc                  32561 non-null uint8
workclass_ Self-emp-not-inc              32561 non-null uint8
workclass_ State-gov                     32561 non-null uint8
workclass_ Without-pay                   32561 non-null uint8
education_ 11th                          32561 non-null uint8
education_ 12th                          32561 non-null uint8
education_ 1st-4th                       32561 non-null uint8
education_ 5th-6th                       32561 non-null uint8
education_ 7th-8th                       32561 non-null uint8
education_ 9th         

In [195]:
#ad_data['marital_status'] = pd.Categorical(ad_data.marital_status)
from sklearn.model_selection import train_test_split

In [196]:
X_train, X_test, y_train,y_test = train_test_split(cdata,target,test_size=0.20,random_state=1234)

In [197]:
X_train.shape, X_test.shape, y_train.shape,y_test.shape

((26048, 59), (6513, 59), (26048,), (6513,))

In [198]:
dt3 = tree.DecisionTreeClassifier()

In [199]:
model4 = dt3.fit(X_train,y_train)


In [200]:
model4.score(X_train,y_train)

0.9999616093366094

In [201]:
np.where(model4.feature_importances_>0.05)

(array([25, 53, 54, 55, 56, 58], dtype=int64),)

In [202]:
colnames = pd.Series(cdata.columns)

In [203]:
imp = pd.Series(np.round(model4.feature_importances_,2))

In [204]:
feat_imp = pd.concat([colnames,imp],axis=1)

In [205]:
feat_imp.columns = ['var','score']

In [206]:
feat_imp.sort_values(by='score', ascending=False).head(10)

var  score
25  marital_status_ Married-civ-spouse   0.20
54                              fnlwgt   0.19
55                       education_num   0.12
53                                 age   0.12
56                        capital_gain   0.11
58                      hours_per_week   0.06
57                        capital_loss   0.04
48                  relationship_ Wife   0.01
42            occupation_ Tech-support   0.01
41                   occupation_ Sales   0.01

In [207]:
imp_features = feat_imp.sort_values(by='score', ascending=False).head(10)

In [219]:
import matplotlib.pyplot as plt
%matplotlib inline


In [216]:
!pip install matplotlib

You are using pip version 9.0.1, however version 19.1.1 is available.
You should consider upgrading via the 'python -m pip install --upgrade pip' command.


In [221]:
model4.score(X_train,y_train)

0.9999616093366094

In [222]:
model4.score(X_test,y_test)

0.8128358667280823

In [ ]:
# steps in K-fold cross validation
#1- split the dataset into K- equal parts
#2- use fold 1 as the testing set and the union of other folds as training set
#3- calculate testing accuracy
#4- repeat step 2 and 3
#5- use the average testing accuracy as the estimate of the out of the sample accuracy

In [ ]:
from sklearn.model_selection import train_test_split, KFold, cross_val_score

In [ ]:
kf = KFold(n_splits=10,shuffle=False)

In [ ]:
kf

In [ ]:
k_range = list(range(1,15))
k_scores = []
for k in k_range:
    dt3 = tree.DecisionTreeClassifier()
    scores = cross_val_score(dt3,X_train,y_train,cv=5,scoring='accuracy')
    k_scores.append(scores.mean())
print(k_scores)


In [ ]:
np.mean(k_scores)

In [ ]:
np.std(k_scores)

In [ ]:
import matplotlib.pyplot as plt
%matplotlib inline

In [ ]:
# plot
plt.plot(k_range,k_scores)


In [ ]:
import seaborn as sns
sns.boxplot(y=k_scores)

In [ ]:
np.round(np.sqrt(59),0)

In [ ]:
YPred = model4.predict(X_test)

In [ ]:
from sklearn.metrics import confusion_matrix,classification_report
print(confusion_matrix(YPred,y_test))
print(classification_report(YPred,y_test))

In [ ]:
# what is the problem with model4:
# biased model as it looks at 6 important features and because it is a single decision tree
# other least important features are not getting a place in the tree construction process

# to remove this bias from the model we can procees with a random forest classifier
# because the RF model strikes a right balance between the bias and variance

# this is done by including the random features and many trees

In [ ]:
import sklearn.ensemble as ske

In [ ]:
################Random Forest Model########################

rf1 = ske.RandomForestClassifier(n_estimators=500, max_features=6).fit(X_train,y_train)
rf1

In [ ]:
#help(ske.RandomForestClassifier())

In [ ]:
YTestProb = rf1.predict_proba(X_test)
YPred = rf1.predict(X_test)

In [ ]:
from sklearn.metrics import confusion_matrix,classification_report
print(confusion_matrix(YPred,y_test))
print(classification_report(YPred,y_test))

In [ ]:
YTestProb = rf1.predict_proba(X_test)
YPred = rf1.predict(X_test)

In [ ]:
from sklearn.metrics import confusion_matrix,classification_report
print(confusion_matrix(YPred,y_test))
print(classification_report(YPred,y_test))

In [ ]:
plt.figure(figsize=(19,10))
pd.DataFrame(rf1.feature_importances_).plot(kind='bar')
plt.show()

In [ ]:
param_grid = {'criterion':['gini','entropy'],
             'max_depth': [2, 5, 7],
             'n_estimators':[100,200,300],
             'max_features': [4,6,8]
             }

In [ ]:
cvRF = ske.RandomForestClassifier()

In [ ]:
cvRF_Param = GridSearch_BestParam(X_train,y_train,cvRF, param_grid,cv=10)

In [ ]:
cvRF_Param

In [ ]:
paramCV = sorted(cvRF_Param,key=itemgetter(1), reverse=True)

In [ ]:
paramCV = paramCV[0].parameters

In [ ]:
paramCV

In [ ]:
cvRF = ske.RandomForestClassifier(n_estimators=paramCV["n_estimators"],
                                  criterion=paramCV["criterion"],
                                  max_depth=paramCV["max_depth"],
                                  max_features=paramCV["max_features"]
                                )

cvRF.fit(X_train,y_train)

In [ ]:
YValidRF = cvRF.predict(X_test)

print (confusion_matrix(y_test,YValidRF))

In [ ]:
cvRF.score(X_train,y_train)

In [ ]:
cvRF.score(X_test,y_test)

In [ ]:
cvRF.feature_importances_

In [ ]:
imp_RF = pd.Series(np.round(cvRF.feature_importances_,2))

In [ ]:
feat_imp['score_rf'] = imp_RF

In [ ]:
feat_imp.sort_values(by='score_rf', ascending=False).head(10)

In [ ]:
feat_imp.sort_values(by='score_rf', ascending=False).tail(10)

In [ ]:
len(np.where(cvRF.feature_importances_<0.01)[0])

In [ ]:
var_list = feat_imp[feat_imp.score_rf<0.01]

In [ ]:
var_list.columns = ['variables', 'score', 'score_rf']

In [ ]:
var_list.shape

In [ ]:
set(var_list.variables)

In [ ]:
cdata_new = cdata[['education_ 11th',
 'education_ 12th',
 'education_ 1st-4th',
 'education_ 5th-6th',
 'education_ 7th-8th',
 'education_ 9th',
 'education_ Assoc-acdm',
 'education_ Assoc-voc',
 'education_ Doctorate',
 'education_ Preschool',
 'education_ Prof-school',
 'education_ Some-college',
 'fnlwgt',
 'marital_status_ Married-AF-spouse',
 'marital_status_ Married-spouse-absent',
 'marital_status_ Separated',
 'marital_status_ Widowed',
 'occupation_ Adm-clerical',
 'occupation_ Armed-Forces',
 'occupation_ Craft-repair',
 'occupation_ Farming-fishing',
 'occupation_ Handlers-cleaners',
 'occupation_ Machine-op-inspct',
 'occupation_ Priv-house-serv',
 'occupation_ Protective-serv',
 'occupation_ Sales',
 'occupation_ Tech-support',
 'occupation_ Transport-moving',
 'race_ Asian-Pac-Islander',
 'race_ Black',
 'race_ Other',
 'race_ White',
 'relationship_ Other-relative',
 'workclass_ Federal-gov',
 'workclass_ Local-gov',
 'workclass_ Never-worked',
 'workclass_ Private',
 'workclass_ Self-emp-inc',
 'workclass_ Self-emp-not-inc',
 'workclass_ State-gov',
 'workclass_ Without-pay']]

In [ ]:
cdata_new.columns

In [ ]:
X_train1, X_test1, y_train1,y_test1 = train_test_split(cdata_new,target,test_size=0.20,random_state=1234)

In [ ]:
X_train1.shape, X_test1.shape, y_train1.shape,y_test1.shape

In [ ]:
dt3.fit(X_train1,y_train1)

In [ ]:
dt3.score(X_train1,y_train1)

In [ ]:
dt3.score(X_test1,y_test1)

In [ ]:
cvRF.fit(X_train1,y_train1)

In [ ]:
cvRF.score(X_train1,y_train1)

In [ ]:
cvRF.score(X_test1,y_test1)

In [ ]:
imp_RF_reducedset = pd.Series(np.round(cvRF.feature_importances_,2)) 

In [ ]:
feat_imp['score_rf_reducedset'] = imp_RF_reducedset

In [ ]:
feat_imp.sort_values(by='score_rf_reducedset', ascending=False).head(10)

# Solution?

In [ ]:
# Boosting Models

In [ ]:
# 1. Gradient Boosting Model
# 2. Xtreme Gradient Boosting
# 3. Stochastic Gradient Boosting 

In [209]:
from sklearn.ensemble import BaggingClassifier
from sklearn.neighbors import KNeighborsClassifier
bagging = BaggingClassifier(KNeighborsClassifier(),max_samples=0.5, max_features=0.5)

In [211]:
bagging.fit(X_train,y_train)

BaggingClassifier(base_estimator=KNeighborsClassifier(algorithm='auto',
                                                      leaf_size=30,
                                                      metric='minkowski',
                                                      metric_params=None,
                                                      n_jobs=None,
                                                      n_neighbors=5, p=2,
                                                      weights='uniform'),
                  bootstrap=True, bootstrap_features=False, max_features=0.5,
                  max_samples=0.5, n_estimators=10, n_jobs=None,
                  oob_score=False, random_state=None, verbose=0,
                  warm_start=False)

In [212]:
bagging.score(X_train,y_train)

0.8244394963144963

In [213]:
bagging.score(X_test,y_test)

0.8088438507600184

In [223]:
from sklearn.ensemble import BaggingClassifier, ExtraTreesClassifier, IsolationForest, VotingClassifier
from sklearn.neighbors import KNeighborsClassifier
bagging = BaggingClassifier(KNeighborsClassifier(),max_samples=0.5, max_features=0.5)

In [224]:
model5 = ExtraTreesClassifier()

In [225]:
model6 = IsolationForest()

In [234]:
eclf = VotingClassifier(estimators=[('dt', model4), ('rf', model6), ('dt1', model5)], voting='hard')

In [228]:
model5.fit(X_train,y_train)

E:\New folder\lib\site-packages\sklearn\ensemble\forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


ExtraTreesClassifier(bootstrap=False, class_weight=None, criterion='gini',
                     max_depth=None, max_features='auto', max_leaf_nodes=None,
                     min_impurity_decrease=0.0, min_impurity_split=None,
                     min_samples_leaf=1, min_samples_split=2,
                     min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=None,
                     oob_score=False, random_state=None, verbose=0,
                     warm_start=False)

In [229]:
model6.fit(X_train,y_train)

E:\New folder\lib\site-packages\sklearn\ensemble\iforest.py:237: FutureWarning: default contamination parameter 0.1 will change in version 0.22 to "auto". This will change the predict method behavior.
  FutureWarning)
E:\New folder\lib\site-packages\sklearn\ensemble\iforest.py:247: FutureWarning: behaviour="old" is deprecated and will be removed in version 0.22. Please use behaviour="new", which makes the decision_function change to match other anomaly detection algorithm API.
  FutureWarning)


IsolationForest(behaviour='old', bootstrap=False, contamination='legacy',
                max_features=1.0, max_samples='auto', n_estimators=100,
                n_jobs=None, random_state=None, verbose=0, warm_start=False)

In [235]:
eclf.fit(X_train,y_train)

E:\New folder\lib\site-packages\sklearn\ensemble\iforest.py:237: FutureWarning: default contamination parameter 0.1 will change in version 0.22 to "auto". This will change the predict method behavior.
  FutureWarning)
E:\New folder\lib\site-packages\sklearn\ensemble\iforest.py:247: FutureWarning: behaviour="old" is deprecated and will be removed in version 0.22. Please use behaviour="new", which makes the decision_function change to match other anomaly detection algorithm API.
  FutureWarning)


VotingClassifier(estimators=[('dt',
                              DecisionTreeClassifier(class_weight=None,
                                                     criterion='gini',
                                                     max_depth=None,
                                                     max_features=None,
                                                     max_leaf_nodes=None,
                                                     min_impurity_decrease=0.0,
                                                     min_impurity_split=None,
                                                     min_samples_leaf=1,
                                                     min_samples_split=2,
                                                     min_weight_fraction_leaf=0.0,
                                                     presort=False,
                                                     random_state=None,
                                                     splitter='best')),
      

In [236]:
model5.score(X_train,y_train)

0.9999616093366094

In [237]:
model5.score(X_test,y_test)

0.8283433133732535

In [242]:
model6.score_samples(X_test)

array([-0.37098033, -0.42978226, -0.51941664, ..., -0.372303  ,
       -0.4134028 , -0.35434111])

# Boosting models

In [248]:
from sklearn.model_selection import cross_val_score
from sklearn.ensemble import AdaBoostClassifier


In [246]:
clf = AdaBoostClassifier(n_estimators=100)
scores = cross_val_score(clf, X_train, y_train, cv=5)
scores.mean()   

0.8657095334407561

In [247]:
scores_test = cross_val_score(clf, X_test, y_test, cv=5)
scores_test.mean() 

0.8530630601954842

In [249]:
from sklearn.ensemble import GradientBoostingClassifier

In [250]:
clf = GradientBoostingClassifier()

In [251]:
scores = cross_val_score(clf, X_train, y_train, cv=5)
scores.mean() 

0.8660550376231304

In [252]:
scores = cross_val_score(clf, X_test, y_test, cv=5)
scores.mean() 

0.8561339600331503

In [253]:
# Xtreme Gradient Boosting model

In [254]:
# stacking models

In [255]:
clf

GradientBoostingClassifier(criterion='friedman_mse', init=None,
                           learning_rate=0.1, loss='deviance', max_depth=3,
                           max_features=None, max_leaf_nodes=None,
                           min_impurity_decrease=0.0, min_impurity_split=None,
                           min_samples_leaf=1, min_samples_split=2,
                           min_weight_fraction_leaf=0.0, n_estimators=100,
                           n_iter_no_change=None, presort='auto',
                           random_state=None, subsample=1.0, tol=0.0001,
                           validation_fraction=0.1, verbose=0,
                           warm_start=False)

In [256]:
help(clf)

Help on GradientBoostingClassifier in module sklearn.ensemble.gradient_boosting object:

class GradientBoostingClassifier(BaseGradientBoosting, sklearn.base.ClassifierMixin)
 |  Gradient Boosting for classification.
 |  
 |  GB builds an additive model in a
 |  forward stage-wise fashion; it allows for the optimization of
 |  arbitrary differentiable loss functions. In each stage ``n_classes_``
 |  regression trees are fit on the negative gradient of the
 |  binomial or multinomial deviance loss function. Binary classification
 |  is a special case where only a single regression tree is induced.
 |  
 |  Read more in the :ref:`User Guide <gradient_boosting>`.
 |  
 |  Parameters
 |  ----------
 |  loss : {'deviance', 'exponential'}, optional (default='deviance')
 |      loss function to be optimized. 'deviance' refers to
 |      deviance (= logistic regression) for classification
 |      with probabilistic outputs. For loss 'exponential' gradient
 |      boosting recovers the AdaBoost 